### ISSUE TITLE
Python: `pulumi.export` fails silently when used in an apply

### ISSUE THREAD

**Issue by jkodroff** [pulumi/pulumi#10782]
When creating a stack output within an apply in Python, no exception is thrown, and the output does not appear.

This program:

```python
import pulumi
import pulumi_command as command

command1 = command.local.Command(
    "command1",
    create="echo \"command1\"",
)


def do_something(arg):
    print("do_something() was called")
    command2 = command.local.Command(
        "command2",
        create="echo \"command2\"",
    )

    try:
        print("Executing export")
        pulumi.export("output2", command2.stdout)
        print("Export did not throw an exception")
    except Exception as e:
        print(f"Export threw an exception: {e}")


pulumi.export("output1", command1.stdout)
command1.stdout.apply(do_something)
```

Produces this output:

```text
Diagnostics:
  pulumi:pulumi:Stack (repro-python-apply-bug-dev):
    do_something() was called
    Executing export
    Export did not throw an exception
 
Outputs:
  + output1: "command1"
```

`pulumi stack output` matches the output above: `command1`'s output is exported, but `command2`'s is not.

**Comment by Frassle** [comment]
So the issue here is one of async races. Stack outputs are registered by the stack code like so:
```
        try:
            func()
        finally:
            self.register_outputs(massage(self.outputs, []))
```
Which means as soon as the main stack function finishes we send off all the outputs we've seen to the engine. But there could be outstanding apply's still running once `func` finishes, which if they are also trying to register outputs will just be missed.

**Comment by jkodroff** [comment]
@Frassle If we have the context to know we are in an apply, even throwing an exception with an actionable error message (don't do this - do it like this instead) would be an improvement over the current behavior.

**Comment by defyjoy** [comment]
This is still outstanding and open . Do we have any solution to this ?

**Comment by Frassle** [comment]
Not really. The only solution we currently have is "don't do this", make sure that `register_outputs` is always called _outside_ the apply statement. `register_outputs` can itself take an output as the value to register, so this should always be possible but I get it's probably not always the most obvious way to write the code.

This (and some other related issues) are all being thought about, but it's likely we'd need a few API change to express these in a way that's easy to use and correct, so might not be doable till 4.0 when we can do a breaking change to fix the interface.

**Comment by jkodroff** [comment]
@Frassle Is there sufficient context to know we're in an apply and at least issue a warning to the user?

**Comment by Frassle** [comment]
There might be actually now we have context vars, we could set a flag on one for when we call into apply and warning if we see that set during register_outputs.

**Comment by justinvp** [comment]
Possibly related to #10637 (need to make Python SDK wait for all applys to run)?

**Comment by zbuchheit** [comment]
Adding some context related to a request that came in and what I experienced in attempting to provide a solution. 

This functionality would be nice for a use case like splitting an array of subnets into individual exports. Exporting each manually can be verbose and tedious to accomplish as a workaround. I would expect the following to work but it silently fails as described in the issue.

`subnet_output.apply(lambda subnets: {pulumi.export(f'subnet-{index}', subnet) for index, subnet in enumerate(subnets)})`
